pkg/member/participant.go

```diff
@@ -0,0 +1,401 @@
+// Copyright 2023 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/election"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/storage/kv"
+	"github.com/tikv/pd/pkg/utils/etcdutil"
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+)
+
+// Participant is used for the election related logic. Compared to its counterpart
+// EmbeddedEtcdMember, Participant relies on etcd for election, but it's decoupled
+// with the embedded etcd. It implements Member interface.
+type Participant struct {
+	leadership *election.Leadership
+	// stored as member type
+	leader atomic.Value
+	client *clientv3.Client
+	// id is unique among all participants
+	id uint64
+	// member is the current participant's info.
+	// TODO: use a more general instead of pdpb.Member. The latter is tightly
+	// coupled with the embedded etcd. For now, we can reuse it and ignore the
+	// fields which are related to embedded etcd, such as PeerUrls and ClientUrls.
+	member     *pdpb.Member
+	rootPath   string
+	leaderPath string
+	// memberValue is the serialized string of `member`. It will be saved in the
+	// eader key when this participant is successfully elected as the leader of
+	// the group. Every write will use it to check the leadership.
+	memberValue string
+}
+
+// NewParticipant create a new Participant.
+func NewParticipant(client *clientv3.Client, id uint64) *Participant {
+	return &Participant{
+		client: client,
+		id:     id,
+	}
+}
+
+// InitInfo initializes the member info. The leader key is path.Join(rootPath, leaderName)
+func (m *Participant) InitInfo(name string, rootPath string, leaderName string, purpose string) {
+	leader := &pdpb.Member{
+		Name:     name,
+		MemberId: m.ID(),
+	}
+
+	data, err := leader.Marshal()
+	if err != nil {
+		// can't fail, so panic here.
+		log.Fatal("marshal leader meet error", zap.Stringer(leaderName, leader), errs.ZapError(errs.ErrMarshalLeader, err))
```

Comment:
```suggestion
		log.Fatal("marshal leader meet error", zap.Stringer(leader-name, leader), errs.ZapError(errs.ErrMarshalLeader, err))
```

Reply:
changed to "tso-leader" in which leader means leader in election

Reply:
can't use "tso-leader" because participant is a more general concept than tso. changed to "leader-name"